Use c10::ThreadPool to send messages

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23968 Use c10::ThreadPool to send messages**

Existing ProcessGroupAgent uses a single thread to send all messages, and
a single thread to listen and process all received messages. This causes
both performance issues and also prevents nested RPCs. For example, when
running nested RPC A->B->A->B, the second recv on B cannot start until
the first recv on B finishes. If the second recv is triggered by a nested
RPC in the first recv, it will deadlock. Ideally, we should expose sth like
responder or FutureResult to the Python land to support nested asynchronous
UDFs.

This diff adds a shared ThreadPool for send and recv. Send use it do send
out messages, and recv use it to process received messages. There is still
a dedicated thread to listen for incoming messages and add it to task queue.
There are two goals: 1) speed up ProcessGroupAgent 2) use ThreadPool as a
temporary solution for (a small number of) nested RPCs

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

mrshenli committed:
Use c10::ThreadPool to send messages

Existing ProcessGroupAgent uses a single thread to send all messages. To speed
up, it now uses a thread pool.

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages. To speed
up, it now uses a thread pool.

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

mrshenli commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -143,48 +137,31 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
-
-    lock.lock();
-  }
+  sendThreadPool_.run(std::bind(
```
Hey @ilia-cher, it seems `ThreadPool::run` made a [copy](https://github.com/pytorch/pytorch/blob/82aecfad6a0a1d7790f7e96157033bcc576d346e/c10/core/thread_pool.h#L57) of the `std::function`. Is there any way to avoid this copy? Copy the bound work here can be expensive.

ilia-cher replied:
it does make a copy to store the task internally, I'm not sure what better approach you'd suggest here, in any way we should demonstrate the issue and the win using a micro-benchmark

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages. To speed
up, it now uses a thread pool.

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

aazzolini commented:
Since the single goal of this PR is speed up, I believe need needs to come along with a micro benchmark. Could you measure the improvement in terms of latency?

ilia-cher reviewed (changes requested):
Second that, perf. related PRs should demonstrate an improvement on a (micro-)benchmark

ilia-cher commented:
also please add more information on what you're trying to achieve here, e.g. I wonder how many new thread pool instances there's going to be in a training process?

mrshenli commented:
@aazzolini @ilia-cher 

I was intended to try out `c10::ThreadPool` on sends before adding it for recvs to unblock @xush6528. Will add some micro-benchmark later. 

@ilia-cher 
There will be one send ThreadPool and one recv ThreadPool per ProcessGroupAgent per process. Is there any concern on this?

BTW, is it true that `c10::ThreadPool` is making a copy of the function and bound args?

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages. To speed
up, it now uses a thread pool.

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

aazzolini commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -143,48 +138,31 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+      };
+
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
```
So the lock here will end up simply making one send execute at a time -- so the only thing that you're parallelizing is the serialization, right?

mrshenli replied:
It's one lock per rank. If this RPC agent communicates with different processes, this would still help, no?

pietern replied:
This changes the ordering semantics of the messages.

E.g. if you want to send 2 messages to the remote, and you queue both of them one after the other without waiting for completion, the execution order is now undefined. Since there are multiple send threads, these messages can be picked up at the same time, and then those threads race to grab the lock.

If those messages are calls that mutate local state on the remote, this can become messy.

mrshenli replied:
> if you want to send 2 messages to the remote, and you queue both of them one after the other without waiting for completion, the execution order is now undefined. 

Will the undefined order be a problem? I thought we are not going to enforce any order on RPC execution? Even if we guarantee FIFO on the caller side, there is still no promise on the order of execution on the callee side, right? Because there might be multiple threads working in parallel serving requests.

> Since there are multiple send threads, these messages can be picked up at the same time, and then those threads race to grab the lock.

Ahh, yes, this will hurt performance.

> If those messages are calls that mutate local state on the remote, this can become messy.

Can this be considered as an application problem? They apps would like to enforce order on mutating some global var, they should add syncs properly?

pietern replied:
> Can this be considered as an application problem? They apps would like to enforce order on mutating some global var, they should add syncs properly?

Yes, we can approach it like this. But from the user point of view it will be unexpected, I think. If you make two calls one after the other and they get reordered, but only if certain timing conditions hit, it will be a PITA to debug. And we can ask ourselves how many users will 1) read the docs, and 2) apply strict ordering. I think it's worth thinking about some kind of ordered mode, which is an optional thing that forces sequencing of RPCs to a single remote. Then in examples we can refer to it, and it will become clear soon enough that by default you shouldn't expect the call order and execution order to be the same.

pritamdamania87 replied:
I don't think we should worry about ordering of RPC calls, most RPC frameworks don't guarantee this. For example, in thrift if you kick off two RPC calls one after the other, there is no guarantee which one will complete first.

@mrshenli Does the thread-safety issue of ProcessGroup apply to all types of ProcessGroups? I was looking at the implementation for ProcessGroupGloo and it seems like the transport layer does have some of its own locking. If send() is not thread-safe for the same tag for ProcessGroupGloo, could you point out which part is not thread-safe?

mrshenli replied:
@pritamdamania87 One of the issue is that, the receiver excepts to see a preamble followed by a payload. If we don't have a lock here, the receiver could get two consecutive preambles.

pritamdamania87 replied:
@mrshenli: Why don't we send the preamble and the payload together in one send call?

aazzolini commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -182,20 +202,11 @@ void ProcessGroupAgent::listenLoop() {
 
     Message message = deserialize(ss);
```
Ideally we'd like to move deserialize to the thread pool as well to unblock the next receive while we're deserializing the previous message. this will allow to increase parallelization (it won't decrease latency of a single call though ...).

mrshenli replied:
agree! we can add the message type into preamble for shutdown. Will make the change.

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -58,12 +58,15 @@ Message deserialize(std::istream& is) {
       pg_(std::move(pg)),
-      nextId_(0) {
+      nextId_(0),
+      sendMutexes_(new std::mutex[pg_->getSize()]),
```
Use `std::vector` instead?

mrshenli replied:
We need dynamic allocation here, and I thought we cannot resize vector of `mutex` because there is no copy/move constructor for `mutex`? Let me try. I also prefer to use `vector` if it is possible.

apaszke replied:
You can use `vector` as long as you never resize it

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -106,11 +102,9 @@ void ProcessGroupAgent::sync() {
-  // Wait until the send queue is depleted.
-  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  // Wait until the all send works are done.
+  // NB: There might be additional send works inserted while waiting.
```
Why is that? Wouldn't `::sync` only be called by the same user thread that ends up issuing the sends?

mrshenli replied:
Yes, but there could be nested RPC calls. Say when calling sync on A, A has one pending send to B, and when running that RPC on B, it triggers B->A->B. In this case one additional send will be inserted into A's task queue.

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -138,53 +132,56 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
-
-  workProduceCV_.notify_one();
+void ProcessGroupAgent::enqueueSend(RpcWork work) {
+  threadPool_.run(std::bind(
```
Nice trick, `std::bind` to force a move capture.

Could you add a note here that this can be changed to use a native move capture when we move to C++14?

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -138,53 +132,56 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+      std::vector<torch::Tensor> payload =
+          {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
+      pg_->send(payload, work.rank_, work.rank_ /* channelTag */)->wait();
+      lock.unlock();
```
You can use a `std::lock_guard` as well, since you don't have to unlock it before it is destructed anyway.

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -138,53 +132,56 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+          futures_.erase(id);
+        }
+      } else {
+        AT_ERROR("unrecognized message type ", work.message_.type());
```
This causes an exception to be thrown, but there is no-one to handle it. If it is really an unrecognized message, shouldn't this be a `TORCH_INTERNAL_ASSERT` and have it crash the process?

mrshenli replied:
How about adding error handling to `FutureMessage` (or just use `ivalue::Future`'s error handle after we make Message an iValue)? So the error will be passed to the user when the sync rpc returns or `FutureMessage::wait()` returns.

pietern replied:
Sounds good. As long as we always have a message to associate this error with, of course. Ideally, this function should be marked noexcept for that reason, to force all errors to propagate back to the user.

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.h:
```diff
@@ -60,12 +58,11 @@ class ProcessGroupAgent : public RpcAgent {
-  std::thread sendThread_;
+  // one mutex per ProcessGroup rank, as ProcessGroup::send is not thread-safe
+  // when using the same tag.
+  std::unique_ptr<std::mutex[]> sendMutexes_;
```
As commented above, I think a `std::vector` is better suited here.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -138,53 +132,56 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
+      std::unique_lock<std::mutex> lock(sendMutexes_.get()[work.rank_]);
+      pg_->send(preamble, work.rank_, work.rank_ /* channelTag */)->wait();
```
Do we need to wait under the lock? This would be pretty expensive since the default timeout for ProcessGroupGloo is 10s.

mrshenli replied:
I assume so, otherwise, the `preamble` and the `payload` tensors could be destructed before the send finish. And that would load to nondeterministic behavior I think.

pritamdamania87 replied:
ProcessGroupGloo returns `SendWork` which captures the tensor to ensure it doesn't get deallocated. We should be able to write something like this without any issues:

```
std::unique_loc<> lock(...)
auto work = pg_->send(....)
lock.unlock()
work.wait()
```

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.h:
```diff
@@ -42,7 +39,8 @@ class ProcessGroupAgent : public RpcAgent {
+  void enqueueSend(RpcWork work);
+  void enqueueRecv(RpcWork work);
   // sending out the message
   void sendLoop();
```
Didn't we remove `sendLoop()` from the cpp file? We do we still have the definition?

mrshenli replied:
Yes, this is removed in my local version.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.h:
```diff
@@ -42,7 +39,8 @@ class ProcessGroupAgent : public RpcAgent {
 
  private:
   // put SendWork into a queue and notify the sendLoop thread
```
Fix this comment since we removed `sendLoop`.

mrshenli commented:
Hey @pritamdamania87 thanks for reviewing. Shall we hold followup reviews on this PR a bit? I have some updates on this one but cannot export it after rebasing to #23569. Will update when #23569 is landed.

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages. To speed
up, it now uses a thread pool.

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages. To speed
up, it now uses a thread pool.

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

mrshenli commented:
Micro-benchmark:

before:

```
test_stress_heavy_rpc (test_rpc.RpcTest) ... 
Rank 3 finished testing heavy_rpc 20 times in 1.841923713684082 seconds.
Rank 0 finished testing heavy_rpc 20 times in 1.8617804050445557 seconds.
Rank 1 finished testing heavy_rpc 20 times in 1.9007983207702637 seconds.
Rank 2 finished testing heavy_rpc 20 times in 1.9105055332183838 seconds.

test_stress_light_rpc (test_rpc.RpcTest) ... 
Rank 2 finished testing light_rpc 1000 times in 1.124621868133545 seconds.
Rank 0 finished testing light_rpc 1000 times in 1.1577978134155273 seconds.
Rank 3 finished testing light_rpc 1000 times in 1.170255422592163 seconds.
Rank 1 finished testing light_rpc 1000 times in 1.1736280918121338 seconds.
```


after (num_send_recv_threads = 4, default):

```
test_stress_heavy_rpc (test_rpc.RpcTest) ... 
Rank 1 finished testing heavy_rpc 20 times in 0.5223677158355713 seconds.
Rank 0 finished testing heavy_rpc 20 times in 0.5939044952392578 seconds.
Rank 2 finished testing heavy_rpc 20 times in 0.6136205196380615 seconds.
Rank 3 finished testing heavy_rpc 20 times in 0.6397840976715088 seconds.

test_stress_light_rpc (test_rpc.RpcTest) ... 
Rank 3 finished testing light_rpc 1000 times in 0.5930540561676025 seconds.
Rank 0 finished testing light_rpc 1000 times in 0.5937092304229736 seconds.
Rank 2 finished testing light_rpc 1000 times in 0.6050472259521484 seconds.
Rank 1 finished testing light_rpc 1000 times in 0.6078634262084961 seconds.
```

after (num_send_recv_threads = 8):

```
test_stress_heavy_rpc (test_rpc.RpcTest) ... 
Rank 1 finished testing heavy_rpc 20 times in 0.34965944290161133 seconds.
Rank 3 finished testing heavy_rpc 20 times in 0.3574059009552002 seconds.
Rank 2 finished testing heavy_rpc 20 times in 0.3725595474243164 seconds.
Rank 0 finished testing heavy_rpc 20 times in 0.41245269775390625 seconds.

test_stress_light_rpc (test_rpc.RpcTest) ... 
Rank 3 finished testing light_rpc 1000 times in 0.5265188217163086 seconds.
Rank 2 finished testing light_rpc 1000 times in 0.5312702655792236 seconds.
Rank 0 finished testing light_rpc 1000 times in 0.5387604236602783 seconds.
Rank 1 finished testing light_rpc 1000 times in 0.541905403137207 seconds.
```

mrshenli commented:
@aazzolini @ilia-cher @pritamdamania87 I think I addressed all the comments above and added a micro-benchmark, could you please help take another look? Thanks!

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -133,13 +131,14 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+    [&](const SendWork& work) {
       std::stringstream ss;
       serialize(work.message_, ss);
       std::string str = ss.str();
```
nit: call this `serialized_payload` instead of just `str`.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -149,36 +148,47 @@ void ProcessGroupAgent::enqueueSend(RpcWork work) {
+      std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
+      pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
+      if (work.message_.type() != MessageType::SHUTDOWN) {
+        std::vector<torch::Tensor> payload ={
```
Why don't we build the payload outside the lock? In the common case, this would avoid extra operations under the lock. In the `shutdown` case, we can build an empty payload.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -187,27 +197,23 @@ void ProcessGroupAgent::enqueueRecv(RpcWork work) {
-    Message message = deserialize(ss);
-
-    if (message.isShutdown()) {
+    if (type == MessageType::SHUTDOWN) {
```
add a LOG here indicating that we are shutting down the listen loop.

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages, and
a single thread to listen and process all received messages. This causes
both performance issues and also prevents nested RPCs. For example, when
running nested RPC A->B->A->B, the second recv on B cannot start until
the first recv on B finishes. If the second recv is triggered by a nested
RPC in the first recv, it will deadlock. Ideally, we should expose sth like
responder or FutureResult to the Python land to support nested asynchronous
UDFs.

This diff adds a shared ThreadPool for send and recv. Send use it do send
out messages, and recv use it to process received messages. There is still
a dedicated thread to listen for incoming messages and add it to task queue.
There are two goals: 1) speed up ProcessGroupAgent 2) use ThreadPool as a
temporary solution for (a small number of) nested RPCs

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages, and
a single thread to listen and process all received messages. This causes
both performance issues and also prevents nested RPCs. For example, when
running nested RPC A->B->A->B, the second recv on B cannot start until
the first recv on B finishes. If the second recv is triggered by a nested
RPC in the first recv, it will deadlock. Ideally, we should expose sth like
responder or FutureResult to the Python land to support nested asynchronous
UDFs.

This diff adds a shared ThreadPool for send and recv. Send use it do send
out messages, and recv use it to process received messages. There is still
a dedicated thread to listen for incoming messages and add it to task queue.
There are two goals: 1) speed up ProcessGroupAgent 2) use ThreadPool as a
temporary solution for (a small number of) nested RPCs

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

mrshenli commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -140,87 +132,99 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
 }
 
 void ProcessGroupAgent::listenLoop() {
+  google::InitGoogleLogging("distributed.rpc");
```
This line helped me get rid of the following warning for the `LOG(INFO)` (Thanks! @pritamdamania87 and @jamarshon):

```
WARNING: Logging before InitGoogleLogging() is written to STDERR
```

But I don't feel right to call it here. I searched a bit, and saw `LOG` is used extensively in `caffe2/` but almost not used at all in `torch/` or `aten/`. @soumith @dzhulgakov @gchanan @ezyang In general, what are the recommended API for logging? Or are we intentionally avoiding that?

ezyang replied:
I think @dzhulgakov knows what the current state here is

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -141,25 +142,32 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
+            )
         };
+        std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
+        pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
```
As I mentioned in my previous comment, I don't think we need to wait under the lock here.

mrshenli committed:
Update on "Use c10::ThreadPool to send messages"


Existing ProcessGroupAgent uses a single thread to send all messages, and
a single thread to listen and process all received messages. This causes
both performance issues and also prevents nested RPCs. For example, when
running nested RPC A->B->A->B, the second recv on B cannot start until
the first recv on B finishes. If the second recv is triggered by a nested
RPC in the first recv, it will deadlock. Ideally, we should expose sth like
responder or FutureResult to the Python land to support nested asynchronous
UDFs.

This diff adds a shared ThreadPool for send and recv. Send use it do send
out messages, and recv use it to process received messages. There is still
a dedicated thread to listen for incoming messages and add it to task queue.
There are two goals: 1) speed up ProcessGroupAgent 2) use ThreadPool as a
temporary solution for (a small number of) nested RPCs

Differential Revision: [D16695091](https://our.internmc.facebook.com/intern/diff/D16695091/)

xush6528 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -140,87 +133,111 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+            pg_->send(payload, work.to_, work.to_ /* channelTag */));
+      }
+      for (auto& pendingSend: pendingSends) {
+        pendingSend->wait();
```
curious about the reason you want to wait for gloo device thread?

mrshenli replied:
Because `work` will be destructed after the lambda function finishes, and the send function itself keep it alive. If destruction happens before send finish, the behavior will be undefined.

xush6528 replied:
@mrshenli 

Since the change you are making here, waiting for device thread to finish the work, is similar to, in network programming, waiting for os kernel to send out the buffer in kernel, making the user code essentially in a blocking fashion.

I wonder is this possible to create the worker object on heap `std::make_shared...` and capture the shared_ptr by value in lambda (adding ref count to the heap object) to keep it alive.

mrshenli replied:
@xush6528 could you please elaborate more? By `worker object` do you mean `SendWork`? If so, why using a `shared_ptr` would help? The `SendWork` is already capture by this lambda. Or are you referring to a different lambda than this one?

mrshenli replied:
Or we could have a separate GC thread for send, and enqueue all async send handles to that thread. The GC thread will then wait on all the send handle in order and delete it when done.

xush6528 replied:
@mrshenli  Sorry, I mean the `ProcessGroup::Work` you are waiting for. Updated the above comment.

mrshenli replied:
> Because work will be destructed after the lambda function finishes, and the send function itself does not keep the tensor alive. If destruction happens before send finish, the behavior will be undefined.

Oh, my bad, when I say `work` above, I actually mean `preamble` and `payload` tensors. We need keep those tensors alive before `send` finishes. And I agree that we can optimize this by capturing `ProcessGroup::Work` and those tensors (all as `std::shared_ptr`s) in a separate lambda, and wait there. Let me add an issue for this later. Thanks!

mrshenli replied:
See #24946
